File: 78.java
import java.util.ArrayList;
import java.util.List;

class solution{
    public static List<List<Integer>> subsets(int[] nums){
        //create a list containing the only empty subset
        List<List<Integer>> ans = new ArrayList<>();
        ans.add(new ArrayList<>()); // add the empty subset

        for (int num : nums){
            int size = ans.size();  // current number of subsets in the ans list
            
            // iterate over all the subsets that EXISTED BEFORE we processed the current 'num'
            for (int i=0; i<size; i++){
                // // get the subset from the existing list
                // List<Integer> existingSubset = ans.get(i);
                // // create a new list (inclusion path)
                // List<Integer> newSubset = new ArrayList<>(existingSubset);
                
                // combining the above 2 steps
                List<Integer> subset = new ArrayList<>(ans.get(i));

                // add the current 'num' to the subset
                subset.add(num);

                // add the subset to 'ans' list
                ans.add(subset);
            }
        }

        return ans;
    }
}

File: 78.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        ans.push_back(vector<int>());

        for (int num : nums) {
            int sz = ans.size();
            for (int i=0; i<sz; i++) {
                vector<int> subset(ans.at(i));
                subset.push_back(num);
                ans.push_back(subset);
            }
        }

        return ans;
    }
};
